pkg/trtl/replica/replica.go

```diff
@@ -0,0 +1,396 @@
+			// local version on this replica.
+			var local *object.Object
+			if local, err = r.db.Object(sync.Object.Key, options.WithNamespace(sync.Object.Namespace)); err != nil {
+				if err == engine.ErrNotFound {
+					// If this is a not found error, then this object exists on the
+					// initiating replica, but not locally, so request a repair.
+					// Note that we have to set the object version to zero to
+					// indicate that we don't have a version of it, so the client
+					// replica's version is later and it sends a repair message.
+					sync.Object.Version = &object.VersionZero
+					sender.Send(sync)
+				} else {
+					// This is an unhandled error, log and return error information.
+					log.Warn().Err(err).
+						Str("namespace", sync.Object.Namespace).
+						Str("key", b64e(sync.Object.Key)).
+						Msg("failed check sync on initiator: could not fetch object meta")
+
+					// If there is an object error, send a response back to the
+					// initiator's CHECK reqest so that it can be logged there as well.
+					sync.Status = replica.Sync_ERROR
+					sync.Error = err.Error()
+					sender.Send(sync)
+				}
+
+				// CHECK is done here if there was an error.
+				continue gossip
+			}
+
+			// Check which version is later, local or remote
+			switch {
+			case local.Version.IsLater(sync.Object.Version):
+				// Send the local object back to the initiating replica as a repair
+				sync.Status = replica.Sync_REPAIR
+				sync.Object = local
+				sync.Error = ""
+				if ok := sender.Send(sync); ok {
+					atomic.AddUint64(&updates, 1)
+				}
+			case sync.Object.Version.IsLater(local.Version):
+				// Send a check request back to the initiating replica to fetch its version
+				sender.Send(sync)
+			default:
+				// The versions are equal, do nothing
+			}
+
+		case replica.Sync_REPAIR:
+			// Receiving a repaired object, check if it's still later than our local
+			// replica's and if so, save it to disk with an update instead of a put.
+			//
+			// NOTE: we must check if it's later than our local version in case a
+			// concurrent Put (a stomp) or concurrent synchronization updated it.
+			// NOTE: honu.Update performs the version checking in a transaction.
+			if err := r.db.Update(sync.Object, options.WithNamespace(sync.Object.Namespace)); err != nil {
+				log.Error().Err(err).
+					Str("namespace", sync.Object.Namespace).
+					Str("key", b64e(sync.Object.Key)).
+					Msg("could not update object from initiating replica")
+				continue gossip
+			}
+			atomic.AddUint64(&repairs, 1)
+
+		case replica.Sync_ERROR:
+			// Something went wrong on the initiator, log and continue
+			if sync.Object != nil {
+				log.Warn().Str("error", sync.Error).
+					Str("key", b64e(sync.Object.Key)).
+					Str("namespace", sync.Object.Namespace).
+					Msg("a replication error occurred")
+			} else {
+				log.Warn().Str("error", sync.Error).Msg("a replication error occurred")
+			}
```

Comment:
Why do we continue if there is an error?

Reply:
This particular error is related to a single object - so if the object has been corrupted, an admin may be able to fix it, but we want to continue with the rest of the synchronization so that all other objects are replicated. If this was a critical error, then instead of sending a `Sync_ERROR`, the initiator would simply close the stream with a gRPC error.